src/ws-server.ts

```diff
@@ -0,0 +1,73 @@
+import { WebSocket } from 'ws';
+import { v4 as uuidv4 } from 'uuid';
+import { WebSocketServer } from 'ws';
+
+import { jsonRpcApi } from './json-rpc-api.js';
+import { createJsonRpcErrorResponse, JsonRpcErrorCodes, JsonRpcResponse } from './lib/json-rpc.js';
+
+export const wsServer = new WebSocketServer({ noServer: true });
+
+wsServer.on('connection', function(socket: WebSocket, _request, _client) {
+  socket['id'] = uuidv4();
+  socket['isAlive'] = true;
+
+  // Pong messages are automatically sent in response to ping messages as required by
+  // the websocket spec. So, no need to send explicit pongs from browser
+  socket.on('pong', function() {
+    this['isAlive'] = true;
+  });
+
+  socket.on('message', async function(dataBuffer) {
+    let dwnRequest;
+
+    try {
+      // deserialize bytes into JSON object
+      dwnRequest = dataBuffer.toString();
+      if (!dwnRequest) {
+        const jsonRpcResponse = createJsonRpcErrorResponse(uuidv4(),
+          JsonRpcErrorCodes.BadRequest, 'request payload required.');
+
+        const responseBuffer = jsonRpcResponseToBuffer(jsonRpcResponse);
+        return socket.send(responseBuffer);
+      }
+
+      dwnRequest = JSON.parse(dwnRequest);
+    } catch(e) {
+      const jsonRpcResponse = createJsonRpcErrorResponse(
+        uuidv4(), JsonRpcErrorCodes.BadRequest, e.message);
+
+      const responseBuffer = jsonRpcResponseToBuffer(jsonRpcResponse);
+      return socket.send(responseBuffer);
+    }
+
+    const { jsonRpcResponse } = await jsonRpcApi.handle(dwnRequest, { transport: 'ws' });
+
+    const responseBuffer = jsonRpcResponseToBuffer(jsonRpcResponse);
+    return socket.send(responseBuffer);
+  });
+});
+
+// Sometimes connections between client <-> server can get borked in such a way that
+// leaves both unaware of the borkage. ping messages can be used as a means to verify
+// that the remote endpoint is still responsive. Server will ping each socket every 30s
+// if a pong hasn't received from a socket by the next ping, the server will terminate
+// the socket connection
+const heartbeatInterval = setInterval(function () {
+  wsServer.clients.forEach(function (socket) {
+    if (socket['isAlive'] === false) {
+      return socket.terminate();
+    }
+
+    socket['isAlive'] = false;
+    socket.ping();
+  });
+}, 30_000);
+
+function jsonRpcResponseToBuffer(jsonRpcResponse: JsonRpcResponse) {
```

Comment:
is this actually needed?  i thought that `ws` supports sending objects?  the [docs](https://github.com/websockets/ws/blob/HEAD/doc/ws.md#websocketsenddata-options-callback) seem to suggest so

Reply:
i'm fine with either way. reason that i went with buffer was because data always arrives as a buffer so i figured: buffer in, buffer out. happy to go with whatever the consensus is. thoughts @frankhinek ?